Resolve relative delete markers on upgrade. Closes #1463

milleruntime opened the pull request:
* Also some minor clean up to Upgrader9to10

ctubbsii commented on server/master/src/main/java/org/apache/accumulo/master/upgrade/Upgrader9to10.java:
```diff
@@ -625,4 +626,39 @@ private static Path resolveRelativePath(String metadataEntry, Key key) {
+    // use Path to check the format is correct
+    Path pathToCheck = new Path(oldDelete);
+    Preconditions.checkState(
+        oldDelete.startsWith("/") && (pathToCheck.depth() == 2 || pathToCheck.depth() == 3),
```
Should have comment to briefly explain intent of this check, so its implementation can be more easily verified. The current comment just says "check the format is correct", but does not explain what it means to be "correct".

milleruntime replied:
Updated comments.  Let me know if it is more clear now

ctubbsii replied:
This is definitely better. Thanks.

ctubbsii commented on server/master/src/main/java/org/apache/accumulo/master/upgrade/Upgrader9to10.java:
```diff
@@ -625,4 +626,39 @@ private static Path resolveRelativePath(String metadataEntry, Key key) {
+
+    // abs path won't be null so return
+    if (pathNoVolume != null)
+      return pathNoVolume;
```
This says "absolute path won't be null", but that's not the same as "relative path must be null". Can relative path be non-null? If so, then this will return for both absolute paths and non-null relative paths.

milleruntime replied:
The ```removeVolume``` method will return null if it doesn't have a volume aka it is a relative path.

keith-turner commented on server/master/src/main/java/org/apache/accumulo/master/upgrade/Upgrader9to10.java:
```diff
@@ -430,26 +432,27 @@ public void upgradeFileDeletes(ServerContext ctx, Ample.DataLevel level) {
-          relPath.getName());
+  static String switchToAllVolumes(Path olddelete) {
+    // for directory, change volume to all volumes
+    if (olddelete.depth() == 4 && !olddelete.getName().startsWith(Constants.BULK_PREFIX)) {
```
Volumes do not have to be depth of 1.  For example a user could configure a volume like `hdfs://localhost:9000/dev/accumulo`.  This would make a dir using that volume have a depth != 4.

milleruntime replied:
Ah OK.  Then we still need to remove the volume before checking.  

@ctubbsii I added a comment above the line and at the beginning of the method.  I will add the ```removeVolume``` call back and update the comments.

keith-turner commented on server/master/src/main/java/org/apache/accumulo/master/upgrade/Upgrader9to10.java:
```diff
@@ -625,4 +626,39 @@ private static Path resolveRelativePath(String metadataEntry, Key key) {
+    Path absPath =
+        new Path(upgradeProperty, VolumeManager.FileType.TABLE.getDirectory() + oldDelete);
+    try {
+      if (!fs.exists(absPath)) {
```
Delete entries do not have to exists.  This can happen when the GC deletes some files, but the GC process dies before it deletes the delete entries.

milleruntime replied:
Good to know, thanks.  I will remove this check.

milleruntime commented:
@ctubbsii I think this is good to go if you are happy with updated the comments

ctubbsii commented on server/master/src/main/java/org/apache/accumulo/master/upgrade/Upgrader9to10.java:
```diff
@@ -625,4 +630,30 @@ private static Path resolveRelativePath(String metadataEntry, Key key) {
+  static Path resolveRelativeDelete(String oldDelete, String upgradeProperty) {
+    Path pathNoVolume = VolumeManager.FileType.TABLE.removeVolume(new Path(oldDelete));
+
+    // removeVolume will return null if path doesn't have a volume aka is a relative path
```
I'm still a bit concerned about the wording of this comment here. Is this an `if` or an `iff (if and only if)`?

milleruntime commented:
I updated the comment again and added test for this scenario to ```FileTypeTest.```  I get a sense that you are more concerned about the FileType method itself then the upgrade code.  There is still potential for the bug you created [here](https://issues.apache.org/jira/browse/ACCUMULO-3577) which I think should be addressed separately.

ctubbsii commented:
The "otherwise" wording makes it clear this is a bi-implication. :smiley_cat:

ctubbsii commented:
To aid future code archaeologists. The references to #1463 in the commits here are a mistake. They should reference #1465 instead. :smiley_cat:
